### DISCUSSION TITLE
Setting signature provider keys and invoking provider sign callbacks

### DISCUSSION THREAD

**Discussion by AbadiNiro** [openssl/openssl#26123] [Q&A]
Hi all, 
Im trying to implement signature provider in order to establish mutual TLS connection and initialize private key inside the provider callbacks during the handshake (for example the CertificateVerify step will invoke my sign callback and inside that I will set my private key first).
I initialize the provider and the required functions in order to make the provider work (signature functions keymgmt functions)

in each function I just print the name of it and return the expected value.

right now i cant reach the sign functions i only get to the keymgmt import and than the handshake failed , i think my problem is in setting up the EVP_PKEY structure.

this is how I load my provider: 

```
OSSL_LIB_CTX * libssl = OSSL_LIB_CTX_new();
OSSL_PROVIDER_add_builtin(libssl ,"custom",my_init);
OSSL_PROVIDER_load(libssl ,"custom");
OSSL_PROVIDER_load(libssl , "default");
EVP_set_default_properties(libssl , "?provider=default");
```

this is how I initialize my SSL_CTX

`SSL_CTX_new_ex(libssl ,"?provider=default" ,TLS_client_method())
`

I created ssl from this ctx

and than I need to set the keys for this ssl instance but i want to set my private key inside the callbacks and not at this stage

```
SSL_use_certificate(ssl, certificate);
SSL_use_PrivateKey(ssl,  ??? );
```

finally I call SSL_connect and it fails at  ClientKeyExchange stage

I tried according to posts here to set the Keys like that :

```
EVP_PKEY *pkey=NULL;
EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_from_name(libssl, "RSA","provider=custom");
EVP_PKEY_fromdata_init(pctx);
EVP_PKEY_fromdata(pctx, &pkey, EVP_PKEY_KEYPAIR, NULL);
```

and pass that pkey to the SSL_use_PrivateKey function , but still fails , i can see that after EVP_PKEY_fromdata function my keymgmt new improt and free callbacks called one by one but the call for set_privateKey fails,

am i missing something inside the providers function or somthing with the key initialization?

**Comment by t8m** [comment]
Does your keymanagement implement has() and match() functions?

**Reply by AbadiNiro** [reply]
Hi thanks a lot for the response,
It is actually was the problem and the function match() was missing.
now im dealing with another problem which is the custom teardown function dont get called , 
I can see that the refcnt is 2 and openssl wont invoke the custom teardown ,

the order of creation and deletion is as follows:
OSSL_LIB_CTX_new //create an empty OSSL_LIB_CTX
OSSL_PROVIDER_add_builtin // added my custom provider
OSSL_PROVIDER_load // load the **default** provider into my OSSL_LIB_CTX which created earlier
OSSL_PROVIDER_load// load the **custom** provider into my OSSL_LIB_CTX which created earlier

in the deletion 
OSSL_PROVIDER_unload // unload the **custom** provider into my OSSL_LIB_CTX which created earlier
OSSL_PROVIDER_unload// unload the **default** provider into my OSSL_LIB_CTX which created earlier
OSSL_LIB_CTX_free // release my LIB_CTX 

i cant understand why my refcnt of the custom provider unload is 2
